libcudacxx/include/cuda/std/__bit/integral.h

```diff
@@ -21,53 +21,140 @@
+  return d;
 }
 
-template <class _Tp>
-_LIBCUDACXX_HIDE_FROM_ABI constexpr enable_if_t<sizeof(_Tp) >= sizeof(uint32_t), _Tp> __ceil2(_Tp __t) noexcept
+_LIBCUDACXX_HIDE_FROM_ABI uint32_t bfind(uint64_t a)
 {
-  return _Tp{1} << (numeric_limits<_Tp>::digits - __countl_zero((_Tp) (__t - 1u)));
+  uint32_t d;
+  asm volatile(
+    "{                                                   \n\t\t"
+    "bfind.u64 %0, %1;                      \n\t\t"
+    "}"
+    : "=r"(d)
+    : "l"(a));
+  return d;
 }
 
-template <class _Tp>
-_LIBCUDACXX_HIDE_FROM_ABI constexpr enable_if_t<sizeof(_Tp) < sizeof(uint32_t), _Tp> __ceil2(_Tp __t) noexcept
+_LIBCUDACXX_HIDE_FROM_ABI uint32_t bfind(unsigned long long a)
 {
-  return (_Tp) ((1u << ((numeric_limits<_Tp>::digits - __countl_zero((_Tp) (__t - 1u)))
-                        + (numeric_limits<unsigned>::digits - numeric_limits<_Tp>::digits)))
-                >> (numeric_limits<unsigned>::digits - numeric_limits<_Tp>::digits));
+  uint32_t d;
+  asm volatile(
+    "{                                                   \n\t\t"
+    "bfind.u64 %0, %1;                      \n\t\t"
+    "}"
+    : "=r"(d)
+    : "l"(a));
+  return d;
 }
 
-template <class _Tp>
-_LIBCUDACXX_HIDE_FROM_ABI constexpr enable_if_t<__cccl_is_unsigned_integer<_Tp>::value, _Tp> bit_floor(_Tp __t) noexcept
+_LIBCUDACXX_HIDE_FROM_ABI uint32_t bfind(__uint128_t a)
 {
-  return __t == 0 ? 0 : static_cast<_Tp>(_Tp{1} << __bit_log2(__t));
+  return 0;
 }
 
+} // namespace cuda::ptx
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+#define _CCCL_CUDA_BUILTIN_ASSUME(...)                           \
+  if (!_CUDA_VSTD::is_constant_evaluated())                      \
+  {                                                              \
+    NV_IF_TARGET(NV_IS_DEVICE, (__builtin_assume(__VA_ARGS__);)) \
+  }
+
 template <class _Tp>
-_LIBCUDACXX_HIDE_FROM_ABI constexpr enable_if_t<__cccl_is_unsigned_integer<_Tp>::value, _Tp> bit_ceil(_Tp __t) noexcept
+_LIBCUDACXX_HIDE_FROM_ABI constexpr _CUDA_VSTD::uint32_t __bit_log2(_Tp __t) noexcept
 {
-  return (__t < 2) ? 1 : static_cast<_Tp>(__ceil2(__t));
+  if (!_CUDA_VSTD::is_constant_evaluated() && sizeof(_Tp) <= 8)
+  {
+    NV_IF_ELSE_TARGET(NV_IS_DEVICE,
+                      (return ::cuda::ptx::bfind(__t);), //
+                      (return numeric_limits<_Tp>::digits - 1 - _CUDA_VSTD::countl_zero(__t);))
+  }
+  else
+  {
+    return numeric_limits<_Tp>::digits - 1 - _CUDA_VSTD::countl_zero(__t);
+  }
 }
 
-template <class _Tp>
-_LIBCUDACXX_HIDE_FROM_ABI constexpr enable_if_t<__cccl_is_unsigned_integer<_Tp>::value, int> bit_width(_Tp __t) noexcept
+_CCCL_TEMPLATE(class _Tp)
+_CCCL_REQUIRES(_CCCL_TRAIT(_CUDA_VSTD::__cccl_is_unsigned_integer, _Tp))
+_CCCL_NODISCARD _LIBCUDACXX_HIDE_FROM_ABI constexpr int bit_width(_Tp __t) noexcept
+{
+  // __bit_log2 returns 0xFFFFFFFF if __t == 0. Since unsigned overflow is well-defined, the result is -1 + 1 = 0
+  using _Up  = _CUDA_VSTD::make_unsigned_t<decltype(+_Tp{})>;
```

Comment:
I dont like that `decltype(+_Tp{})` if we need a well defined integer type we should make that explicit and not rely on integer promotion